diesel/src/pg/expression/functions.rs

```diff
@@ -1634,3 +1636,83 @@ define_sql_function! {
         text_array: Arr,
     ) -> Arr::Out;
 }
+
+#[cfg(feature = "postgres_backend")]
+define_sql_function! {
+    /// Returns the number of elements in the top-level JSON array
+    ///
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     #[cfg(feature = "serde_json")]
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # #[cfg(feature = "serde_json")]
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::json_array_length;
+    /// #     use serde_json::json;
+    /// #     use diesel::sql_types::{Integer, Json};
+    /// #     let connection = &mut establish_connection();
+    /// let result = diesel::select(json_array_length::<Json, _>(json!([1, 2, 3])))
+    ///     .get_result::<i32>(connection)?;
+    ///
+    /// assert_eq!(result, 3);
+    ///
+    /// let result = diesel::select(json_array_length::<Json, _>(json!([])))
+    ///     .get_result::<i32>(connection)?;
+    ///
+    /// assert_eq!(result, 0);
+    ///
+    ///
+    ///
+    /// #     Ok(())
+    /// # }
+    /// ```
+
+    fn json_array_length<E: JsonOrNullableJson + SingleValue>(e: E) -> Integer;
```

Comment:
Sorry for taking a few days to write this answer, this somehow got lost between all the other PR's :disappointed: 

This is already mostly the correct signature. We cannot restrict this function to only accept json arrays at compile time as that's a purely runtime concept, so it's fine to accept `JsonOrNullableJson` here. What needs to be done is to make the nullability of the return value dependend on the input nullability, which means if you put in a `Nullable<Json>` we want to return a `Nullable<Integer>` as the function returns `null` for `null` inputs.

You can do that by changing the function signature as follows:

```suggestion
    fn json_array_length<E: JsonOrNullableJson + MaybeNullableValue<Integer>>(e: E) -> E::Out;
```

As for the test: You should be able to test for null values like this:

```
let result = diesel::select(json_array_length::<Nullable<Json>, _>(None::<serde_json::Value>)).get_result::<Option<i32>(conn);
```

(The same applies to the jsonb variant of this function)
```

Reply:
Thank you so much for the explanation! I have pushed in a new commit for the required changed. Please let me know if anything else needs to be changed/fixed. Thanks once again!